File: Scala/graph_clustering.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

import scala.collection.{SortedMap, mutable}
import scala.util.control.Breaks._
import scala.collection.mutable.Map
import scala.collection.mutable.Queue
import scala.collection.mutable.ListBuffer
import java.io.{File, PrintWriter}


object piyush_chaudhari_task1 {
  def main(args : Array[String]) : Unit = {
    val conf1 = new SparkConf().setAppName("Betweenness").setMaster("local[*]")
    val sc1 = new SparkContext(conf1)
    val filterThreshold = args(0).toInt
    val yelpDataFile = args(1)
    val betweennessOutputFile = args(2)
    val communityOutputFile = args(3)
//    val yelpDataFile = "C:\\Users\\cpiyu\\Downloads\\INF 553 Data Mining\\Assignments\\Assignment 4\\sample_data.csv"
    val yelpData = sc1.textFile(yelpDataFile)
    val yelpRDD = yelpData.map(x => x.split(",")).filter(row => row(0) != "user_id").map(x => ((x(0), List(x(1))))).reduceByKey((x, y) => x ::: y).collectAsMap()
    val allUsers = yelpRDD.keys.toList
    //    print("Temp : "+temp)

    var combinations = allUsers.combinations(2)
    //    print("Comb : "+combinations.length)
    //    val loop = new Breaks;
    var allEdges = ListBuffer.empty[(String, String)]
    for (comb <- combinations) {
      var firstBusinessSet = yelpRDD(comb(0)).toSet
      var secondBusinessSet = yelpRDD(comb(1)).toSet
      var intersection = firstBusinessSet.intersect(secondBusinessSet)
      var firstEdge = Tuple2(comb(0), comb(1))
      var secondEdge = Tuple2(comb(1), comb(0))
      if (intersection.size >= filterThreshold) {
        allEdges = allEdges :+ firstEdge
        allEdges = allEdges :+ secondEdge
      }
    }
//    print("all edges : "+allEdges.size)
    var adjacencyList = Map[String, List[String]]().withDefaultValue(List())
    for (edge <- allEdges) {
      adjacencyList(edge._1) = edge._2 :: adjacencyList(edge._1)
    }

    var adjacencyListUpdated = Map[String, ListBuffer[String]]().withDefaultValue(ListBuffer())
//    for (edge <- allEdges) {
//      println("Node : "+(edge._1,edge._2))
//      println("Init : "+adjacencyListUpdated(edge._1).size)
//      adjacencyListUpdated(edge._1) = adjacencyListUpdated(edge._1) += edge._2
//      println("Final : "+adjacencyListUpdated(edge._1).size)
//    }

    var count2 = 0
    for(node <- adjacencyList.keys){
      adjacencyListUpdated(node) = adjacencyList(node).to[ListBuffer]
      count2 += adjacencyList(node).size
    }
    var count1 = 0
    for(node <- adjacencyListUpdated.keys){
      count1 += adjacencyListUpdated(node).size
    }
//    println("Nodes : " + adjacencyListUpdated.keys.size)
//    println("Original edges : " + count2)
//    println("New edges : "+count1)

    val users = adjacencyList.keys

    def GN(start: String, graph: Map[String, ListBuffer[String]]): Map[(String, String), Double] = {
      var edgeBetweenness = Map.empty[(String, String), Double].withDefaultValue(0)
      var head = start
      var explored = ListBuffer[String]()
      var queue = Queue[String]()
      queue += start
      var bfsTree = ListBuffer[String]()
      var levels = Map.empty[String, Int]
      var parentDict = Map.empty[String, List[String]].withDefaultValue(List())
      var treeLevel = Map.empty[String, Double].withDefaultValue(0)
      var children = ListBuffer[String]()
      //      levels(start) = 0
      levels += start -> 0
      //      treeLevel(head) = 1
      treeLevel += head -> 1
      //      print("queue",queue)
      while (queue.isEmpty == false) {
        val node = queue.dequeue()
        //        print("node : "+node)
        explored += node
        //        print("explored : ",explored)
        children = graph(node)

        //        print("Node : "+node+" Children : "+children)
        for (child <- children) {
          if (!explored.contains(child)) {
            if (!queue.contains(child)) {
              bfsTree += child
              queue += child
              levels(child) = levels(node) + 1
            }
            if (levels(child) != levels(node)) {
              //println("Node : "+node)
              //              println("Init : ",parentDict(child).size)
              parentDict(child) = node :: parentDict(child)
              //              println("Final : ",parentDict(child).size)
              treeLevel(child) += treeLevel(node)
            }
          }
        }
      }

      //print("bfs size : "+bfsTree.size)
      var nodeDict = Map.empty[String, Double].withDefaultValue(1)
      for (i <- bfsTree.size - 1 to 0 by -1) {
        var addition:Double = 0
        var currentNode = bfsTree(i)
        if (parentDict.contains(currentNode)) {
          var parents = parentDict(currentNode)
          for (parent <- parents) {
            //            println("parent : "+treeLevel(parent))
            //            println("curr_level : "+treeLevel(currentNode))
            //            println("curr_node : "+nodeDict(currentNode))

            addition = ((treeLevel(parent) / treeLevel(currentNode)) * nodeDict(currentNode))
            //            println("init : "+nodeDict(parent))
            nodeDict(parent) += addition
            //            println("addition : "+addition)
            //            var edge = (currentNode,parent)
            var edge = (List(currentNode, parent).sortWith(_ < _))
            var t = edge match {
              case List(a, b) => (a, b)
            }
            //            print("Edge : "+edge+" Betwenness : "+addition)
            edgeBetweenness(t) = addition
          }
        }
      }
      //      print(edgeBetweenness)
      return edgeBetweenness

    }

      def returnBetweenness(graph:Map[String, ListBuffer[String]]): Map[(String, String), Double] ={
        var combinedBetweenness = Map.empty[(String, String), Double].withDefaultValue(0)
        var finalEdgeBetweenness = Map.empty[(String, String), Double]
        var count = 0
        for (user <- users) {
          var singleNodeBetweenness = GN(user, graph)

          //      println("user : "+user)
          //println("Intermediate edges : "+singleNodeBetweenness.keys.size)

          combinedBetweenness = combinedBetweenness ++ singleNodeBetweenness.map { case (k, v) => k -> (v + combinedBetweenness.getOrElse(k, 0.toDouble)) }
          //      combinedBetweenness = (combinedBetweenness.keySet ++ singleNodeBetweenness.keySet).map {i=> (i,combinedBetweenness.getOrElse(i,0.0f) + singleNodeBetweenness.getOrElse(i,0.0f))}.toMap


          //      }
          singleNodeBetweenness.clear()
        }
        finalEdgeBetweenness = combinedBetweenness
        //    print("final : ", combinedBetweenness.toSeq.sortBy(-_._2))
        //    println("size : ", combinedBetweenness.keys.size)
        for((k,v) <- finalEdgeBetweenness){
          finalEdgeBetweenness(k) = v/2
        }
        var a = finalEdgeBetweenness.toList.sortBy(-_._2).toMap
        return finalEdgeBetweenness
      }
//    print(a)
//        print("final : ", finalEdgeBetweenness.toSeq.sortBy(-_._2))
    //    var a = sc1.parallelize(allUsers).map(x => GN(x,adjacencyList))

    var aValuesDict = Map.empty[(String, String), Double]
    var edgeCount = (allEdges.size / 2).toDouble
//    println("edges : "+edgeCount)
    var nodeDegreeDict = Map.empty[String, Double]
    var modularity:Double = 0
    var A:Double = 0
//    var difference = 0.0f
    for (node <- users) {
      nodeDegreeDict(node) = adjacencyListUpdated(node).size
    }
    for (initialNode <- users) {
      for (finalNode <- users) {
        if (adjacencyListUpdated(initialNode).contains(finalNode)) {
          A = 1
        }
        else {
          A = 0
        }
        var difference = (A - ((nodeDegreeDict(initialNode) * nodeDegreeDict(finalNode)) / (2 * edgeCount))) / (2 * edgeCount)
        var edge = (initialNode, finalNode)
        aValuesDict(edge) = A
        modularity += difference
      }
    }

//    var count = 0
//    for((k,v) <- aValuesDict){
//      if(v == 1){
//        count += 1
//      }
//    }
//    print("Count : "+count)
//    println("Null modularity : " + modularity)

    def DFS(presentConnectedComponent: ListBuffer[String], v: String, visited: ListBuffer[String], graph: Map[String, ListBuffer[String]]): ListBuffer[String] = {
      var tempVisited = ListBuffer.empty[String]
      var tempConnectedComponents = ListBuffer.empty[String]
      tempVisited = visited
      tempConnectedComponents = presentConnectedComponent
//      println("Node : "+v)
//      println("init visited : "+tempVisited.size)
      tempVisited += v
//      println("updated visited: "+tempVisited.size)
//      println("init cc: "+tempConnectedComponents.size)
      tempConnectedComponents += v
//      println("updated cc: "+tempConnectedComponents.size)
//      println()
//      println()
      for (u <- graph(v)) {
        if (!tempVisited.contains(u)) {
          tempConnectedComponents = DFS(tempConnectedComponents, u, tempVisited, graph)
        }
      }
      return tempConnectedComponents
    }

    def connectedComponents(graph: Map[String, ListBuffer[String]]): ListBuffer[ListBuffer[String]] = {
      var visited = ListBuffer.empty[String]
      var allConnectedComponents = ListBuffer.empty[ListBuffer[String]]
      for (i <- graph.keys) {
        if (!visited.contains(i)) {
          var temp = ListBuffer.empty[String]
          //          var ret = DFS(temp, i, visited, graph)
          allConnectedComponents.append(DFS(temp, i, visited, graph))
//          println("cc length : "+allConnectedComponents)
        }
      }
      return allConnectedComponents
    }

    /*
    var aValuesDict = Map.empty[(String,String),Double]
    var edgeCount = allEdges.size/2.toDouble
    var nodeDegreeDict = Map.empty[String,Double]
    */
    var initialConnectedComponents = ListBuffer.empty[ListBuffer[String]]
    initialConnectedComponents = connectedComponents(adjacencyListUpdated)
    var firstModularity = modularity
//    println("Connected components : "+initialConnectedComponents.size)
    var modelConfigiration = (modularity,initialConnectedComponents)

    def calculateModularity(nodeDegreeDict: Map[String, Double], community: ListBuffer[String], edgeCount: Double, aValuesDict:Map[(String, String), Double]): Double = {
      var communityModularity:Double = 0
      var A:Double = 0
      for (initialNode <- community) {
        for (finalNode <- community) {
          A = aValuesDict((initialNode,finalNode))
          var difference = (A - (nodeDegreeDict(initialNode) * nodeDegreeDict(finalNode) / (2 * edgeCount))) / (2 * edgeCount)
          communityModularity += difference
        }
      }
      return communityModularity
    }

    def swap(originalDict:Map[(String, String), Double]):mutable.LinkedHashMap[Double,ListBuffer[(String,String)]] = {
//      var returnBetweennessDict = mutable.LinkedHashMap.empty[Double,ListBuffer[(String,String)]].withDefaultValue(ListBuffer())
      var reverseBetweennessDict = Map.empty[Double,List[(String,String)]].withDefaultValue(List())
      var tempReverseBetweennessDict = Map.empty[Double,ListBuffer[(String,String)]].withDefaultValue(ListBuffer())
      var test = Map.empty[Double,(String,String)].withDefaultValue((" "," "))
      for((k,v) <- originalDict){
//        print("Initial : "+reverseBetweennessDict(v).size)
//        println("Key : "+reverseBetweennessDict(v))
//        reverseBetweennessDict(v).append(k)
        reverseBetweennessDict(v) = k :: reverseBetweennessDict(v)
//        print(", final : "+reverseBetweennessDict(v).size)
//        println()
      }
      for(node <- reverseBetweennessDict.keys){
        tempReverseBetweennessDict(node) = reverseBetweennessDict(node).to[ListBuffer]

      }
      var returnBetweennessDict = mutable.LinkedHashMap(tempReverseBetweennessDict.toSeq.sortBy(-_._1):_*)
      return returnBetweennessDict
    }

    var finalEdgeBetweenness = returnBetweenness(adjacencyListUpdated)

    val betweennessToWrite = mutable.LinkedHashMap(finalEdgeBetweenness.toSeq.sortBy(_._1).sortBy(-_._2):_*)
    val file = new PrintWriter(new File(betweennessOutputFile))
//    file.write("hi")
    //    val bw = new BufferedWriter(new FileWriter(file))
    var length = betweennessToWrite.keys.size
    var counter = 1
    for((k,v) <- betweennessToWrite){
//      file.write("hello")
//      print("key : "+k+" value : "+v)
      file.write("('")
      file.write(k._1.toString)
      file.write("', '")
      file.write(k._2.toString)
//      file.write(k.toString())
      file.write("')")
      file.write(", ")
      file.write(v.toString())
      if(counter < length){
        file.write("\n")
      }
      counter += 1
    }
    file.close()
//    println("Write done")

//    print("final betweenness : "+finalEdgeBetweenness)
    var returnBetweennessDict = swap(finalEdgeBetweenness)
//    var d = mutable.LinkedHashMap(c.toSeq.sortBy(-_._1):_*)
//    var reverseBetweennessDict  = collection.mutable.Map(c.toSeq: _*)


//    println("final : "+finalEdgeBetweenness)
//    println("final2 : "+finalEdgeBetweenness.valuesIterator.max)
//    println("original_! : "+b.head._2)
//    println("reversed : "+returnBetweennessDict)


    breakable{
      while (true){
        var temp = returnBetweennessDict
        var a = temp.keys.head
//        print("key : "+a)

//        var count1 = 0
//        for(node <- adjacencyListUpdated.keys){
//          count1 += adjacencyListUpdated(node).size
//        }

//        println("Initial length : "+count1)
        for(edge <- returnBetweennessDict(a)){
//          println("Removing edge : "+edge+" with betweenness : "+ a)
          adjacencyListUpdated(edge._1) -= edge._2
          adjacencyListUpdated(edge._2) -= edge._1
        }

//        var count2 = 0
//        for(node <- adjacencyListUpdated.keys){
//          count2 += adjacencyListUpdated(node).size
//        }
//        println("Final length : "+count2)

        var newModularity:Double = 0
        var communites = ListBuffer.empty[ListBuffer[String]]
        communites = connectedComponents(adjacencyListUpdated)
        for(community <- communites){
          newModularity += calculateModularity(nodeDegreeDict,community,edgeCount,aValuesDict)
        }

        modularity = newModularity
//        println("Connected Comp : "+initialConnectedComponents.size+" Mod : "+modularity+" Old Mod : "+firstModularity)
        if(modularity > firstModularity){
           firstModularity = modularity
           initialConnectedComponents = communites

        }
        finalEdgeBetweenness = returnBetweenness(adjacencyListUpdated)
        returnBetweennessDict.clear()
        var updatedBetweennessDict = swap(finalEdgeBetweenness)
        returnBetweennessDict = updatedBetweennessDict

        var count = 0
        for((k,v) <- adjacencyListUpdated){
          count += v.size
        }
        if(count == 0){
          break()
        }
      }

    }
//    println("Size : "+initialConnectedComponents.size)

    var finalConnectedComponents = ListBuffer.empty[ListBuffer[String]]
    for (component <- initialConnectedComponents){
      var community = component.sorted
      finalConnectedComponents.append(community)
    }
    def sort[A : Ordering](coll: Seq[Iterable[A]]) = coll.sorted

    var a = sort(finalConnectedComponents)
    var sortedConnectedComponents = a.sortWith(_.size < _.size)
//    println("a : "+sortedConnectedComponents)
    val communityFile = new PrintWriter(new File(communityOutputFile))
    var communityLength = sortedConnectedComponents.size
    var communityCount = 1
    for(community <- sortedConnectedComponents){
//      println()

      var length = community.size
      var count = 1
      for(component <- community){
        communityFile.write("'")
        communityFile.write(component.toString)
        communityFile.write("'")
        if(count < length){
          communityFile.write(", ")
        }
        count += 1
      }
      if(communityCount < communityLength){
        communityFile.write("\n")
      }
      communityCount += 1
    }
    communityFile.close()
    sc1.stop()
  }
}

File: Python/graph_clustering.py
import pyspark
import itertools
from itertools import combinations
from pyspark import SparkContext
import csv
import sys

# sc1 = SparkContext('local[*]','',conf=conf)
# conf = pyspark.SparkConf()
# conf.set("spark.executor.heartbeatInterval","5s")
# conf.set("spark.network.timeout","10s")
sc = SparkContext("local[*]", "GN")

# yelpFile = "C:\\Users\\cpiyu\\Downloads\\INF 553 Data Mining\\Assignments\\Assignment 4\\sample_data.csv"
filterThreshold = int(sys.argv[1])
yelpFile = sys.argv[2]
betweennessOutputFile = sys.argv[3]
communityOutputFile = sys.argv[4]
yelpData = sc.textFile(yelpFile)
yelpDataRDD = yelpData.map(lambda x:x.split(",")).filter(lambda x:str(x[0]) != 'user_id').map(lambda x: (x[0],[x[1]]))
groupedUsers = yelpDataRDD.reduceByKey(lambda x,y:x+y).collectAsMap()
# distinctUsers = yelpDataRDD.map(lambda x:x[0]).distinct()
allUsers = groupedUsers.keys()
allEdges = list()
comb = combinations(allUsers,2)
for c in list(comb):
  firstBusinessSet = set()
  secondBusinessSet = set()
  firstBusinessSet = set(groupedUsers[c[0]])
  secondBusinessSet = set(groupedUsers[c[1]])
  if(len(firstBusinessSet.intersection(secondBusinessSet)) >= filterThreshold):
    allEdges.append((c[0],c[1]))
    allEdges.append((c[1],c[0]))

# print("all edges : ",len(allUsers))

adjacencyList = dict()
for edge in allEdges:
  adjacencyList.setdefault(edge[0],[])
  adjacencyList[edge[0]] += [edge[1]]

# print("Number of nodes : ",len(adjacencyList.keys()))
filteredUsers = adjacencyList.keys()
initialAdjacencyList = adjacencyList


def GirvanNewmann(graph, start):

  edgeBetweenness = dict()
  # keep track of all visited nodes
  head = start
  explored = []
  # keep track of nodes to be checked
  queue = [start]
  bfsTree = list()
  bfsTree.append(start)
  parentDict = dict()
  levels = dict()
  levels[start] = 0
  treeLevel = {head:1}
  while queue:
      node = queue.pop(0)
      explored.append(node)
      children = graph[node]
      for child in children:
          if child not in explored:
              if(child not in queue):
                bfsTree.append(child)
                queue.append(child)
                levels[child]= levels[node]+1
              parentDict.setdefault(child,[])
              treeLevel.setdefault(child,0)
              if(levels[child] != levels[node]):  
                parentDict[child] += [node]
                treeLevel[child] += treeLevel[node]
              # print(treeLevel)
      # print(parentDict)  
      # treeLevel.setdefault(node,0)
      # treeLevel[node] += treeLevel[parentDict[node]]     
  # print("Parent : ",parentDict)
  # print("SP : ",treeLevel)
  # print(levels)
  # print(bfsTree)
  # print("bfs length : ",len(bfsTree))
  nodeDict = dict()
  
  for i in range(len(bfsTree) - 1,-1,-1):
    currentNode = bfsTree[i]
    nodeDict.setdefault(currentNode,1)
    if(currentNode in parentDict):  
      parents = parentDict[currentNode]
      for parent in parents:
        nodeDict.setdefault(parent,1)
        # addition = float(nodeDict[currentNode]/len(parents))
        addition = float((treeLevel[parent]/treeLevel[currentNode]) * nodeDict[currentNode])
        nodeDict[parent] += addition
        #edge = (min(currentNode,parent),max(currentNode,parent))
        edge = tuple(sorted(tuple((currentNode,parent))))
        edgeBetweenness.setdefault(edge,0)
        edgeBetweenness[edge] = addition
  return edgeBetweenness
  # print("len : ",len(edgeBetweenness.keys()))
  # return edgeBetweenness
  # return (edge,edgeBetweenness[edge])

def returnBetweenness(a):
  combinedBetweenness = dict()
  finalEdgeBetweenness = dict()
  for j,i in enumerate(filteredUsers):
    singleNodeBetweenness = dict()
    singleNodeBetweenness = GirvanNewmann(a,i)
    if(j == 0):
      combinedBetweenness = singleNodeBetweenness.copy()
      # print("2 : ",combinedBetweenness)
      # print("1 : ",singleNodeBetweenness)
    else:
      # print("Every iteration length : ",len(singleNodeBetweenness.keys()))
      # print("2 : ",combinedBetweenness)
      # print("1 : ",singleNodeBetweenness)
      combinedBetweenness = {x: combinedBetweenness.get(x, 0) + singleNodeBetweenness.get(x, 0) for x in set(combinedBetweenness).union(singleNodeBetweenness)}

      # combinedBetweenness = combinedBetweenness.updatetemp1
  finalEdgeBetweenness = combinedBetweenness
  # print("Size : ",len(combinedBetweenness.keys()))
  for k, v in finalEdgeBetweenness.items():
    finalEdgeBetweenness[k] = float(v/2)
  finalEdgeBetweenness = dict(sorted(finalEdgeBetweenness.items(),key = lambda x:x[0][1]))
  finalEdgeBetweenness = dict(sorted(finalEdgeBetweenness.items(),key = lambda x:x[0][0]))
  finalEdgeBetweenness = dict(sorted(finalEdgeBetweenness.items(),key = lambda x:x[1],reverse=True))
  return finalEdgeBetweenness

finalEdgeBetweenness = returnBetweenness(adjacencyList)
adjacencyListBackup = adjacencyList
# print(finalEdgeBetweenness)
# finalEdgeBetweenness = dict()
# for i in filteredUsers:
#   GirvanNewmann(adjacencyList,i)
# print("Size : ",len(edgeBetweenness.keys()))
# for k, v in edgeBetweenness.items():
#   finalEdgeBetweenness[k] = float(sum(v)/2)
  # print(k, sum(v))

# finalEdgeBetweenness = dict(sorted(finalEdgeBetweenness.items(),key = lambda x:x[1],reverse=True))
edge = ()
aValuesDict = dict()
edgeCount = len(allEdges)/2
nodeDegreeDict = dict()
for node in filteredUsers:
  nodeDegreeDict[node] = len(adjacencyList[node])
modularity = 0
for initialNode in filteredUsers:
  for finalNode in filteredUsers:
    if(finalNode in adjacencyList[initialNode]):
      A = 1
    else:
      A = 0
    difference = (A - 0.5*nodeDegreeDict[initialNode]*nodeDegreeDict[finalNode]/edgeCount)/(2*edgeCount)
    # edge = tuple(sorted((initialNode,finalNode)))
    # aValuesDict[edge] = A
    aValuesDict[(initialNode,finalNode)] = A
    modularity += difference
# print("Null modularity : ",modularity)

def DFS(presentConnectedComponent,v,visited,graph):
  visited.append(v)
  presentConnectedComponent.append(v)
  # print(temp)
  for u in graph[v]:
    if(u not in visited):
      # print("u",u)
      presentConnectedComponent = DFS(presentConnectedComponent,u,visited,graph)
  return presentConnectedComponent    

def connectedComponents(graph):
  visited = list()
  allConnectedComponents = list()
  for i in graph.keys():
    # print(i)
    if(i not in visited):
      temp = []
      allConnectedComponents.append(DFS(temp,i,visited,graph))
  return allConnectedComponents

initialConnectedComponents = connectedComponents(adjacencyList)
modelConfiguration = (modularity,initialConnectedComponents)  

# def calculateModularity(adjacencyList,community,edgeCount,aValuesDict):
#   # print("edge : ",edgeCount)
#   nodeDegreeDict = dict()
#   accessed = list()
#   for node in community:
#     nodeDegreeDict[node] = len(adjacencyList[node])
#   modularity = 0
#   edge = ()
#   for initialNode in community:
#     for finalNode in community:
#       edge = tuple(sorted((initialNode,finalNode)))
#       if(edge not in accessed):
#         A = aValuesDict[edge]
#         difference = (A - 0.5*nodeDegreeDict[initialNode]*nodeDegreeDict[finalNode]/edgeCount)/(2*edgeCount)
#         modularity += difference
#         accessed.append(edge)
#         edge = ()  
#   return modularity

def calculateModularity(nodeDegreeDict,community,edgeCount,aValuesDict):
  
  modularity = 0
  for initialNode in community:
    for finalNode in community:
      A = aValuesDict[(initialNode,finalNode)]
      difference = (A - 0.5*nodeDegreeDict[initialNode]*nodeDegreeDict[finalNode]/edgeCount)/(2*edgeCount)
      modularity += difference  
  return modularity

mod = 0

# for community in initialConnectedComponents:
#   mod += calculateModularity(adjacencyList,community,edgeCount,aValuesDict)
# modelConfiguration = (modularity,initialConnectedComponents)
# print("init : ",mod)

def swap(b):
  reverseBetweennessDict = dict()
  for k,v in b.items():
    reverseBetweennessDict.setdefault(v,[])
    reverseBetweennessDict[v] += [k]
  return reverseBetweennessDict
# print(reverseBetweennessDict)
reverseBetweennessDict = swap(finalEdgeBetweenness)
# print("initial : ",reverseBetweennessDict)

i = 0
while(1):
# for val in reverseBetweennessDict.items():
  # print("key : ",val)
  
  # print("rev : ",reverseBetweennessDict)
  
  temp = reverseBetweennessDict
  for edge in reverseBetweennessDict[list(temp.keys())[0]]:
    # print("Iteration no : ",i," ,Removing edge : ",edge," with betweenness value : ",list(temp.keys())[0])
    adjacencyList[edge[0]].remove(edge[1])
    adjacencyList[edge[1]].remove(edge[0])
  # print("\n")
  i = i+1
    # l = 0
    # for i in adjacencyList.keys():
    #   l += len(adjacencyList[i])
    # print("len : ",l)
  newModularity = 0
  communities = connectedComponents(adjacencyList)
    
  for community in communities:
    newModularity += calculateModularity(nodeDegreeDict,community,edgeCount,aValuesDict)
  modularity = newModularity
  # print("Modularity after removing ",modularity)
  if(modularity > modelConfiguration[0]):
    modelConfiguration = (modularity,communities)
  # print("final : ",modelConfiguration)

  l = 0
  # for i in adjacencyList.keys():
  #   l += len(adjacencyList[i])
  # print("len : ",l)  
  finalEdgeBetweennessDict = returnBetweenness(adjacencyList)
  # print("updated dict : ",finalEdgeBetweennessDict)
  reverseBetweennessDict.clear()
  updatedBetweennessDict = dict()
  updatedBetweennessDict = swap(finalEdgeBetweennessDict) 
  reverseBetweennessDict = updatedBetweennessDict.copy()
  # print("Comm : ",len(communities), ". Modularity : ",modularity)
  # print("final : ",reverseBetweennessDict)

  count = 0
  for k,v in adjacencyList.items():
    count += len(v)
  if(count == 0):
    break

# print(len(modelConfiguration[1]))
# print(modelConfiguration[1])
finalCommunities = modelConfiguration[1]

for value in finalCommunities:
  value.sort()
asd = sorted(finalCommunities)
# finalCommunities.sort(key = len)
finalCommunities = sorted(asd,key=len)
# finalCommunities = sorted(,key=len)
# print(finalCommunities)
# print("Maximum modularity : ",modelConfiguration[0])

temp1 = ""
fo = open(betweennessOutputFile,'w')
for k,v in finalEdgeBetweenness.items():
  temp1 += str(k) + ', '+ str(v) + '\n'
fo.write(temp1.rstrip())
fo.close()

temp = ""
with open(communityOutputFile,'w',newline='') as f:
  # csv_writer = csv.writer(f)
  # csv_writer.writerows(finalCommunities)
  for value in finalCommunities:
    temp += str(value)[1:-1] + "\n"
  f.write(temp.rstrip())
f.close()
# print(finalCommunities)

sc.stop()
